app/cdap/components/DirectiveInput/index.tsx

```diff
@@ -0,0 +1,263 @@
+  color: #94ec98;
+  font-size: 14px;
+  margin-bottom: 0;
+`;
+
+const DirectiveBox = styled(Box)`
+  background-color: #ffffff;
+  position: fixed;
+  bottom: 93px;
+  width: 100%;
+`;
+
+const CloseIconButton = styled(IconButton)`
+  color: #ffffff;
+`;
+
+export default function({
+  columnNamesList,
+  onDirectiveInputHandler,
+  onClose,
+  openDirectivePanel,
+}: IDirectiveInputProps) {
+  const [directiveInput, dispatch] = useReducer(reducer, initialDirectiveInputState);
+  enum IDirectiveActions {
+    INPUT_DIRECTIVE,
+    DIRECTIVE_SET,
+    APPLIED_DIRECTIVE,
+    DIRECTIVE_COLUMN_COUNT,
+    DIRECTIVE_USAGE_LIST,
+    DIRECTIVE_LIST,
+    DIRECTIVE_PASTE,
+  }
+  const {
+    inputDirective,
+    isDirectiveSet,
+    directivesList,
+    appliedDirective,
+    directiveColumnCount,
+    isDirectivePaste,
+    directiveUsageList,
+  } = directiveInput;
+  const [directiveColumns, setDirectiveColumns] = useState([]);
+  const [enterCount, setEnterCount] = useState(0);
+  const directiveRef = useRef();
+
+  const handleDirectiveChange = (value) => {
+    if (!value) {
+      setEnterCount(0);
+      dispatch({
+        type: IDirectiveActions.INPUT_DIRECTIVE,
+        payload: false,
+      });
+    }
+    const inputText = value.split(' ');
+    if (directivesList.filter((el) => el.directive == inputText[0]).length) {
+      if (TWO_COLUMN_DIRECTIVE.includes(inputText[0])) {
+        dispatch({
+          type: IDirectiveActions.DIRECTIVE_COLUMN_COUNT,
+          payload: 2,
+        });
+      } else if (MULTIPLE_COLUMN_DIRECTIVE.includes(inputText[0])) {
+        dispatch({
+          type: IDirectiveActions.DIRECTIVE_COLUMN_COUNT,
+          payload: 0,
+        });
+      }
+      dispatch({
+        type: IDirectiveActions.DIRECTIVE_SET,
+        payload: true,
+      });
+      dispatch({
+        type: IDirectiveActions.APPLIED_DIRECTIVE,
+        payload: [inputText[0]],
+      });
+    }
+    if (isDirectiveSet) {
+      columnNamesList.forEach((column) => {
+        if (value.indexOf(column.label) !== -1) {
+          setDirectiveColumns((prev) =>
+            !prev.includes(column.label) ? [...prev, column.label] : prev
+          );
+        } else {
+          setDirectiveColumns((prev) => prev.filter((el) => el != column.label));
+          setEnterCount(appliedDirective.length + directiveColumns.length);
+        }
+      });
+    }
+    dispatch({
+      type: IDirectiveActions.INPUT_DIRECTIVE,
+      payload: value,
+    });
+  };
+
+  useEffect(() => {
+    dispatch({
+      type: IDirectiveActions.DIRECTIVE_USAGE_LIST,
+      payload: formatDirectiveUsageData(inputDirective, directivesList),
+    });
+  }, [inputDirective]);
+
+  const handleKeyDownEvent = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.keyCode === 86) {
+      dispatch({
+        type: IDirectiveActions.DIRECTIVE_PASTE,
+        payload: true,
+      });
+    } else if (event.key === 'Enter') {
+      setEnterCount((prev) => {
```

Comment:
a lot of this can be done from inside useReducer - also you have access to all of the state inside of use reducer so you can move all of this state logic there

Reply:
Done